### PR TITLE
506 feature request arpaname tool

### PR DESCRIPTION
**Pull request by Lost-Rain**
Adds a new tool called "Arpaname Tool" for translating IP addresses (IPv4 and IPv6) to their corresponding IN-ADDR.ARPA or IP6.ARPA names. This tool includes form validation, command execution, and output display functionalities. Simplified IP validation and separated into IPv4 & IPv6.

### CODE REVIEW

File: src/components/Arpaname/arpaname.tsx [hunk 1]
```diff
@@ -0,0 +1,108 @@
+import { Button, LoadingOverlay, Stack, TextInput, Title } from "@mantine/core";
+import { useForm } from "@mantine/form";
+import { useCallback, useState } from "react";
+import { CommandHelper } from "../../utils/CommandHelper";
+import ConsoleWrapper from "../ConsoleWrapper/ConsoleWrapper";
+import { UserGuide } from "../UserGuide/UserGuide";
+import { SaveOutputToTextFile } from "../SaveOutputToFile/SaveOutputToTextFile";
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:7]: - [ ] Upgrade to V2

File: src/components/Arpaname/arpaname.tsx [hunk 2]
```diff
@@ -0,0 +1,108 @@
+import { Button, LoadingOverlay, Stack, TextInput, Title } from "@mantine/core";
+import { useForm } from "@mantine/form";
+import { useCallback, useState } from "react";
+import { CommandHelper } from "../../utils/CommandHelper";
+import ConsoleWrapper from "../ConsoleWrapper/ConsoleWrapper";
+import { UserGuide } from "../UserGuide/UserGuide";
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:6]: - [ ] Upgrade to render component

File: src/components/Arpaname/arpaname.tsx [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import { SaveOutputToTextFile } from "../SaveOutputToFile/SaveOutputToTextFile";
+import { LoadingOverlayAndCancelButton } from "../OverlayAndCancelButton/OverlayAndCancelButton";
+
+const title = "Arpaname Tool";
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:10]: - [ ] move inside function definition.
- [ ] Remove 'Tool'

File: src/components/Arpaname/arpaname.tsx [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+const description_userguide =
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:11]: - [ ] move inside function definition.
- [ ] split description and guide into two distinct constants

File: src/components/Arpaname/arpaname.tsx [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    "Arpaname translates IP addresses (IPv4 and IPv6) to the corresponding IN-ADDR.ARPA or IP6.ARPA names.";
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:12]: - [ ] to their corresponding

File: src/components/Arpaname/arpaname.tsx [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+interface FormValuesType {
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:14]: - [ ] Add docstrings

File: src/components/Arpaname/arpaname.tsx [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    ipAddress: string;
+}
+
+const ArpanameTool = () => {
+    const [loading, setLoading] = useState(false);
+    const [output, setOutput] = useState("");
+    const [pid, setPid] = useState("");
+    const [errorMessage, setErrorMessage] = useState("");
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:22]: - [ ] Remove this functionality and append using the callbackfunctions from the helper method.

File: src/components/Arpaname/arpaname.tsx [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+    let form = useForm<FormValuesType>({
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:24]: - [ ] Doc string

File: src/components/Arpaname/arpaname.tsx [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+const title = "Arpaname Tool";
+const description_userguide =
+    "Arpaname translates IP addresses (IPv4 and IPv6) to the corresponding IN-ADDR.ARPA or IP6.ARPA names.";
+
+interface FormValuesType {
+    ipAddress: string;
+}
+
+const ArpanameTool = () => {
+    const [loading, setLoading] = useState(false);
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:19]: - [ ] Add component constant section and comment
- [ ] Add component state variable section and comment

File: src/components/Arpaname/arpaname.tsx [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    const [pid, setPid] = useState("");
+    const [errorMessage, setErrorMessage] = useState("");
+
+    let form = useForm<FormValuesType>({
+        initialValues: {
+            ipAddress: "",
+        },
+    });
+
+    const handleProcessData = useCallback((data: string) => {
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:30]: - [ ] doc string
   /**
     * handleProcessData: Callback to handle and append new data from the child process to the output.
     * It updates the state by appending the new data received to the existing output.
     * @param {string} data - The data received from the child process.
     */

File: src/components/Arpaname/arpaname.tsx [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+        setOutput((prevOutput) => prevOutput + "\n" + data);
+    }, []);
+
+    const handleProcessTermination = useCallback(
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:34]: - [ ] Doc String

    /**
     * handleProcessTermination: Callback to handle the termination of the child process.
     * Once the process termination is handled, it clears the process PID reference and
     * deactivates the loading overlay.
     * @param {object} param - An object containing information about the process termination.
     * @param {number} param.code - The exit code of the terminated process.
     * @param {number} param.signal - The signal code indicating how the process was terminated.
     */

File: src/components/Arpaname/arpaname.tsx [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+            } else {
+                handleProcessData(`\nProcess terminated with exit code: ${code} and signal code: ${signal}`);
+            }
+            setPid("");
+            setLoading(false);
+        },
+        [handleProcessData]
+    );
+
+    const validateIPAddress = (ip: string) => {
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:49]: - [ ] Doc String

File: src/components/Arpaname/arpaname.tsx [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+            setPid(result.pid);
+        } catch (e: any) {
+            setOutput(`An error occurred: ${e.message}`);
+            setPid("");
+        } finally {
+            setLoading(false);
+        }
+    };
+
+    const clearOutput = useCallback(() => {
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:83]: - [ ] doc string

File: src/components/Arpaname/arpaname.tsx [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+
+    const validateIPAddress = (ip: string) => {
+        // IPv4 pattern
+        const ipv4Pattern = /^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;
+        // IPv6 pattern
+        const ipv6Pattern = /^([0-9a-fA-F]{1,4}:){7}([0-9a-fA-F]{1,4}|:)$/;
+        return ipv4Pattern.test(ip) || ipv6Pattern.test(ip);
+    };
+
+    const onSubmit = async (values: FormValuesType) => {
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:57]: - [ ] Doc String

File: src/components/Arpaname/arpaname.tsx [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+        if (!validateIPAddress(values.ipAddress)) {
+            setOutput("The input is not a valid IP address. Please try again.");
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:59]: - [ ] do not call the setOutput outside of its handle function. It is okay in the clear function when we do not want to append

File: src/components/Arpaname/arpaname.tsx [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+        }
+    };
+
+    const clearOutput = useCallback(() => {
+        setOutput("");
+        setErrorMessage("");
+    }, [setOutput, setErrorMessage]);
+
+    return (
+        <form onSubmit={form.onSubmit(onSubmit)}>
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:89]: - [ ] Add Render component
- [ ] Add dependency checker at runtime.

File: src/components/Arpaname/arpaname.tsx [hunk 17]
```diff
@@ -0,0 +1,108 @@
+import { Button, LoadingOverlay, Stack, TextInput, Title } from "@mantine/core";
+import { useForm } from "@mantine/form";
+import { useCallback, useState } from "react";
+import { CommandHelper } from "../../utils/CommandHelper";
+import ConsoleWrapper from "../ConsoleWrapper/ConsoleWrapper";
+import { UserGuide } from "../UserGuide/UserGuide";
+import { SaveOutputToTextFile } from "../SaveOutputToFile/SaveOutputToTextFile";
+import { LoadingOverlayAndCancelButton } from "../OverlayAndCancelButton/OverlayAndCancelButton";
```
Comment (bradleybogg) [inline src/components/Arpaname/arpaname.tsx:8]: - [ ] Add dependency checker

**Review by bradleybogg — CHANGES_REQUESTED** [review]
Hi @Lost-Rain 

This is a good start but it doesnt conform to our baselining standards.
Please review existing components that have completed baselining such as AirBase for reference.

Thank you!

File: src/components/Arpaname/arpaname.tsx [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+                    feature_description={description}
+                    dependencies={dependencies}
+                ></InstallationModal>
+            )}
+            <form onSubmit={form.onSubmit((values) => onSubmit(values))}>
+                <Stack>
+                    {LoadingOverlayAndCancelButton(loading, pidTarget)}
+                    <TextInput label={"IP address"} required {...form.getInputProps("ipAddress")}></TextInput>
+                    <Button type={"submit"}>Lookup</Button>
+                    <ConsoleWrapper output={output} clearOutputCallback={clearOutput} />
```
Comment (michaeljpigott) [inline src/components/Arpaname/arpaname.tsx:148]: This is where the SaveOutputToTextfile_v2 component should be placed.

File: src/components/RouteWrapper.tsx [hunk 19]
```diff
@@ -284,6 +285,13 @@ export const ROUTES: RouteProperties[] = [
         description: "Arjun can find query parameters for URL endpoints.",
         category: "Web Application Testing",
     },
+    {
+        name: "Arpaname",
+        path: "/tools/Arpaname",
+        element: <ArpanameTool />,
+        description: "Perform reverse DNS lookups for IP addresses.",
```
Comment (michaeljpigott) [inline src/components/RouteWrapper.tsx:292]: Please make the description consistent with the description constant in arpaname.tsx.

File: src-tauri/tauri.conf.json [hunk 20]
```diff
@@ -18,13 +18,13 @@
...
                 "scope": [
                     {
-                        "name": "pkexec",
-                        "cmd": "pkexec",
+                        "name": "arpaname",
+                        "cmd": "arpaname",
                         "args": true
                     },
                     {
-                        "name": "which",
```
Comment (michaeljpigott) [inline src-tauri/tauri.conf.json:26]: Why did you remove the which command? Please leave because I am not sure what other tools require that command.

**Review by michaeljpigott — CHANGES_REQUESTED** [review]
Hi @Lost-Rain 

You are getting closer. Please add the SaveOutputToTextFile_v2 functionality to arpaname, use runCommandGetPidAndOuput instead of Pkexec (root privileges are not required), and restore the which command to tauri.conf.json. 

Kind regards

Michael

**Review by michaeljpigott — CHANGES_REQUESTED** [review]
Hi @Lost-Rain 

Thank you for making those changes. 

I am finding that the InstallationModal is popping up even though I have installed Arpaname via Bind9. Not sure how to fix this problem because it is unusual that a dependency does not equate to a command:
image

This might be a fix for future work. If you can add the variables for steps, tutorial and sourcelink in RenderComponent, then we should be done. 

Kind regards

Michael

**Review by michaeljpigott — APPROVED** [review]

**Pull request merged**